src/spaceObjects/spaceship.cpp

```diff
@@ -295,9 +295,24 @@ void SpaceShip::applyTemplateValues()
     energy_level = max_energy_level = ship_template->energy_storage_amount;
 
     impulse_max_speed = ship_template->impulse_speed;
-    impulse_max_reverse_speed = ship_template->impulse_reverse_speed;
+    if(ship_template->impulse_reverse_speed != -1)
+    {
+        impulse_max_reverse_speed = ship_template->impulse_reverse_speed;
```

Comment:
Does it mean, that if I use the `StarShip:setSpeed()` it will only modify Forward impulse speed (even though I didn't change reverse speed)? 

If so, I would propose changes to have these functions exported to LUA: 
- `StarShip:setSpeed(speed)` => `impulse_max_reverse_speed = impulse_max_speed = speed`
- `StarShip:setForwardSpeed(speed)` => `impulse_max_speed = speed`
- `StarShip:setReverseSpeed(speed)` => `impulse_max_reverse_speed = speed`

That way it will keep the same functionality (and honestly, it would make sense). 

It we merge this commit, mission script like this will work differently (and strangely):
```ps = PlayerShip():setTemplate("Atlantis")  -- code from template: atlantis_variant:setSpeed(90, 10, 20)
ps:setSpeed(120)```
This would set Atlantis forward speed to 120, but it will keep reverse on 90. And this WILL be confusing to mission authors. They did not change Reverse speed, but the speeds are different. And it could break some mission scripts of theirs. 

DTTO on ship_acceleration.

Reply:
Waiting for acceptance or not of std::optional so SetSpeed could just be amended.
For the getter, not fond of deprecated methods but I can't still decide if having a x,y = getImpulseMaxSpeed() which would break formatter strings or multi affectation like x,y,z = getImpulseMaxSpeed(),10 is really worse. 
Maybe an optional getImpulseMaxSpeed(front) (by default) and getImpulseMaxSpeed(rear) or a getImpulseMaxSpeeds() ?
I will come with additional proposal.

Reply:
For now, for me optional arguments solves the setter issues
**On shiptemplate**
setSpeed have a variable number of args.
In lua, 

- setSpeed(forward speed, turn rate, forward acceleration) will set all speed and all acceleration to the requested value

- setSpeed(forward speed, turn rate, forward acceleration, reverse speed) will set forward speed and reverse speed as required, acceleration is the same in both way (don't think we always want to set different acceleration...)

- setSpeed(forward speed, turn rate, forward acceleration, reverse speed; reverse acceleration) will set forward speed, reverse speed, acceleration and reverse acceleration as required

Please note there is no reverse turn rate, I don't think its useful but if added, this would mess logic of argument order.

** On spaceship **
- setMaxImpulseSpeed(speed) will set max forward speed and reverse speed with the same value
- setMaxImpulseSpeed(speed, reverse speed) will set max forward speed and max reverse speed as required

Remains getMaxImpulseSpeed
Here are the proposals :

1. getImpulseMaxSpeed() will return 2 values, like getVelocity 
Pros : 
- you can discard second value
-  it will not break this kind of scenarion : x = getImpulseMaxSpeed() considering forward speed is the wanted speed.
Cons : 
- will break this kind of scenario : x,t = getImpulseMaxSpeed(),OtherFunctionReturningT() (also in string.format("%f %f %f",getImpulseMaxSpeed(), OtherFunctionReturningT())

2. getImpulseMaxSpeed() will return only forward speed. getImpulseMaxSpeed(forward) will return forward speed. getImpulseMaxSpeed(reverse) will return reverse speed.
Pros :
- Won't break anything (except for scripts which won't take reverse speed into account...)
Cons :
- Ugly
- Not so simple to understand why getImpulseMaxSpeed would be forward by default... 

3. deprecated getImpulseMaxSpeed() which returns forward speed, or maybe null if reverse speed is different.
As hemmond proposed, but with only one method for the new function.
getImpulseMaxSpeeds() which will return both speed (forward and reverse)
Pros :
- Won't break anything (except for scripts which won't take reverse speed into account...)
Cons :
- Deprecated function

Reply:
I agree with your setters, I think it is future-proof enough. And messing up logic in ship template methods (if this case happens in the future) is smaller pain than messing it up in functions that are used throughout the mission scripts (and therefore read a lot more by a lot more people). 

About your proposals about getters: 

1. Option 1 is no-go, unless Daid says otherwise. It will break missions. 
2. I can life with option 2 with selectors "forward", "reverse", "both" and empty (which means "forward"). Description why it is not so simple to understand can be explained in documentation (for backwards compatibility, please do not use it in the future). 
3. I agree in principle with option 3, but one letter difference will confuse a lot of mission creators. It is easy to miss that letter and then be confused why it returns different number of values than expected. 
4. My proposal (option number 4): 

Add methods `getImpulseMaxForwardSpeed()`, `getImpulseMaxReverseSpeed()`, `getImpulseMaxBothSpeeds()` and deprecate function `getImpulseMaxSpeed()`. I know `getImpulseMaxBothSpeeds()` is ugly, but if you drop it, you fall to the same trap as in option 3 - confusion when missing a letter.

Reply:
Went with 3 with your amelioration proposal on name.